File: Python/problem_036.py
"""
Problem 36 - Double-base palindromes

The decimal number, 585 = 1001001001 (binary), is palindromic in both bases.

Find the sum of all numbers, less than one million, which are palindromic in
base 10 and base 2.

(Please note that the palindromic number, in either base, may not include
leading zeros.)
"""
from itertools import count, takewhile


def palindromes():
    """Yields all base 10 palindromes"""
    for n in count(1):
        if str(n) == str(n)[::-1]:
            yield n


def is_palindrome_2(n):
    b = bin(n)[2:]

    return b == b[::-1]


def solution():
    return sum(n for n in takewhile(lambda a: a < 1e6, palindromes())
               if is_palindrome_2(n))


if __name__ == '__main__':
    print(solution())

File: go/problem_036.go
// Problem 36 - Double-base palindromes
//
// The decimal number, 585 = 1001001001 (binary), is palindromic in both bases.
//
// Find the sum of all numbers, less than one million, which are palindromic in
// base 10 and base 2.
//
// (Please note that the palindromic number, in either base, may not include
// leading zeros.)
package main

import (
	"fmt"
	"math"
	"strconv"

	"./common"
)

func main() {
	fmt.Println(solution())
}

func solution() (sum int) {
	for _, n := range palindromesUpToLength(6) {
		if common.IsPalindrome(fmt.Sprintf("%b", n)) {
			sum += n
		}
	}

	return
}

func palindromesUpToLength(maxLength int) []int {
	count := int(math.Pow10((maxLength+1)/2)) - 1
	if maxLength%2 == 0 {
		count *= 2
	} else {
		count += count / 10
	}

	palindromes := make([]int, count)
	for ix, length := 0, 1; length <= maxLength; length++ {
		start := int(math.Pow10((length - 1) / 2))
		limit := start * 10

		for n := start; n < limit; n, ix = n+1, ix+1 {
			if length%2 == 1 {
				palindromes[ix] = n*start + reverse(n/10)
			} else {
				palindromes[ix] = n*limit + reverse(n)
			}
		}
	}

	return palindromes
}

func reverse(n int) int {
	reversed, _ := strconv.Atoi(common.ReverseString(strconv.Itoa(n)))
	return reversed
}
